stdlib/iter.mli

```diff
@@ -0,0 +1,42 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*                 Simon Cruanes                                          *)
+(*                                                                        *)
+(*   Copyright 2001 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(* Module [Iter]: functional iterators *)
+
+(** @since NEXT_RELEASE *)
+
+type ('a, 's) step =
+  | Done
+  | Skip of 's
+  | Yield of 'a * 's
+
+type +_ t =
+  | Sequence : 's * ('s -> ('a,'s) step) -> 'a t
+
+val empty : 'a t
+
+val return : 'a -> 'a t
+
+val map : ('a -> 'b) -> 'a t -> 'b t
+
+val filter : ('a -> bool) -> 'a t -> 'a t
+
+val filter_map : ('a -> 'b option) -> 'a t -> 'b t
+
+val flat_map : ('a -> 'b t) -> 'a t -> 'b t
+
+val fold_left : ('a -> 'b -> 'a) -> 'a -> 'b t -> 'a
```

Comment:
@paurkedal, `fold: ('b -> 'a -> 'b) -> 'b -> 'a t -> 'b` behaves just as well with nesting:

```ocaml
fold (fold f) : 'b -> 'a t t -> 'b
```

Plus, I would protest it is often nicer with partial evaluation:

```ocaml
let sum = fold (+) 0
let of_seq = fold add empty
```
In particular, `seq |> fold f init` is as a common pattern as `init |> fold f seq`.

However, your signature being the one adopted everywhere else in the stdlib makes your case stronger.